### ISSUE TITLE
Experiment on String-reverse dataset with Transformers

### ISSUE THREAD

**Issue by vfdev-5** [pytorch-ignite/trainite#6]
Find limitations of Transformer architecture.
For a fixed sized transformer (num_layers, num_heads), train on string-reverse dataset for 2 options:
    - in distribution: large strings in train / val sets
    - out of distribution: small strings in train and large on val sets
We can iterate over num_layers, num_heads to make a table of where transformer is working 100% good and where it starts failing.

The outcome of the experiment can be a blog post showing the usage of trainite and it's usefulness in a small research project. 

EDIT:

A) We want to find the seq length values where fixed transformers starts decreasing its perfs

Fixed tokenizer: a-z,0-9 -> int

We do 10 experiments where we iterate over max seq length, for i in range(10)
- pick fixed transformers: num layers = 2, embed_dim = 32, num heads = 1, posencoding = absolute
- dataset train: fixed_length: 16 * i, alphabet: a-p
- dataset val: fixed_length: 16 * i, alphabet: p-z
- Perfs: train accuracy and val accuracy

We plot a graph for the fixed transformer:
X: fixed_length
Y: train acc, val acc

B)

**Comment by TahaZahid05** [comment]
B) Experiment with using RoPE as positional encoding. Performing prior experiments done on absolute positional encoding.

**Comment by TahaZahid05** [comment]
Results of RoPE encoding on the same experiment we did last week:
Train on: 1-124
Test on: sweep lengths between 1-512

Do the above for different model capcities, Layers=[1,2,4], Heads=[1,2,4]. 
1) RoPE:

<img width="3570" height="2070" alt="Image" src="https://github.com/user-attachments/assets/9f979cff-8ad2-492f-86df-3ee28fb747f1" />


2) Absolute positional encoding for reference:

<img width="3570" height="2070" alt="Image" src="https://github.com/user-attachments/assets/6e179e80-2772-4ba5-a51d-bde2a92b2c72" />

**Comment by TahaZahid05** [comment]
Now, RoPE on the experiment that Victor asked us to do.

Fixed transformer model I used:
hidden_size: 64
num_layers: 2
num_heads: 2

Trained on fixed length sweep of 16 * i where i in range [1,11] for alphabets a-p
Val on fixed length sweep of 16 * i where i in range [1,11] for alphabets q-z

Results:

1) Accuracy graphs for both Train and Val as training went on:

<img width="6417" height="2690" alt="Image" src="https://github.com/user-attachments/assets/88850a8a-e6a1-4c87-94a0-e2145e1aee75" />

2) Overall accuracy results:

<img width="2969" height="1770" alt="Image" src="https://github.com/user-attachments/assets/1af85fed-3068-469b-b8a4-5636f47e85fb" />

3) Loss curves for all runs

<img width="6416" height="2690" alt="Image" src="https://github.com/user-attachments/assets/712c8dbe-1e0c-4ffc-9cea-5368130cee40" />

Validation accuracy stayed at 0% across all experiments as expected, since q-z embeddings were not developed so they stayed randomly initialised, and hence zero results on them.

From the loss curves, we can see that training accuracy went to 100% when the loss plateaued to 0. In case of seq_length = 32 (and 80), the loss platued but not all the way to 0 which led to 0% training accuracy. For higher seq_len, it seems the model needs more epochs or more capacity to get better train accuracy as loss hasn't fully plateaued.

We also do not know whether the model memorised the training set or actually learned it. To test that out, we need accuracy on a hold-out dataset on alphabets p-z (for the same fixed length) to see if it generalises.

**Comment by aaishwarymishra** [comment]
<img width="2084" height="1475" alt="Image" src="https://github.com/user-attachments/assets/79ec7558-4602-4fc3-8030-cefd4647e70b" />

<img width="2084" height="1475" alt="Image" src="https://github.com/user-attachments/assets/09d2fc26-7ac1-4753-9e58-ccee0d33e76d" />

<img width="2684" height="740" alt="Image" src="https://github.com/user-attachments/assets/7ad9808c-a9be-466b-8f6d-8d49ad524b7e" />

Here , train and val results for fixed sized transformer of layers = 4, heads = 2, embedding_dim = 128, feed forward_dim = 512.

With training dataset size of 7k, val 1.5K

I noticed training model on a-p and then validating on p-z will results in wrong results as model has never seen p-z so the embedding values for them are random.

I went through different approach what I did , I trained model on universal characters, alphabets, numbers and punctuations.

I then created the validation dataset from those same character with same seed, but different unique strings not in  dataset for validation.

The accuracy spiked between 50-75 character length out of no where not sure why.

I also experimented with other configs too.

**Comment by aaishwarymishra** [comment]
<img width="2084" height="1475" alt="Image" src="https://github.com/user-attachments/assets/80986134-3b4d-480c-8791-a2b8b7e95fd4" />

For original plan a-p for train and p-z for val.

**Comment by vfdev-5** [comment]
@aaishwarymishra once we are done with what we have in the code base we can close this issue

**Comment by aaishwarymishra** [comment]
closing in the favour of #115

**Comment by vfdev-5** [comment]
- https://github.com/pytorch-ignite/trainite/tree/main/examples/string_reversal
- https://pytorch-ignite.ai/blog/string-reversal-example-trainite/
